### ISSUE TITLE
Secondary rate limits: getting how long to wait

### ISSUE THREAD

**Issue by programLyrique** [XAMPPRocky/octocrab#349]
I sometimes get hit by secondary rate limits when searching code and I would like to know how long exactly I should wait before retrying (I am doing exponential backoff currently...).
According to the [github documentation](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#secondary-rate-limits), I should try after `retry-after` seconds if this is the response header, or the `x-ratelimit-reset` header.

Is there a way to retrieve that in octocrab? 
I would have expected to find it in the `errors` field of `GithubError` but `errors` is always `None` in the case of secondary rate limits for me.

**Comment by XAMPPRocky** [comment]
Thank you for your issue!

Personally I don't think this should be something that users have to worry about for the most part. It's an implementation detail of the client. Our current handling of errors should also do this for you.

**Comment by jrozner** [comment]
I've also been running into this not just with Octocrab but also with the official Ruby library. I think there is something specific with the search API that is different.

**Comment by programLyrique** [comment]
There is actually specific rate limits for search: https://docs.github.com/en/rest/search?apiVersion=2022-11-28#about-search

**Comment by programLyrique** [comment]
It seems that `ratelimit` endpoint just shows the primary rate limits, not the secondary ones (and I don't think Github provides any more information than the `x-ratelimit` about the secondary ones).

Here is a request I do immediatly after getting hit by a secondary rate limit error (and just before another one, as my exponential backoff has not reached a long enough duration):

```rust
RateLimit { resources: Resources { core: Rate { limit: 5000, used: 0, remaining: 5000, reset: 1682344037 }, 
search: Rate { limit: 30, used: 0, remaining: 30, reset: 1682340497 }, 
graphql: Some(Rate { limit: 5000, used: 0, remaining: 5000, reset: 1682344037 }), integration_manifest: Some(Rate { limit: 5000, used: 0, remaining: 5000, reset: 1682344037 }), 
scim: Some(Rate { limit: 15000, used: 0, remaining: 15000, reset: 1682344037 }), 
source_import: Some(Rate { limit: 100, used: 0, remaining: 100, reset: 1682340497 }), 
code_scanning_upload: Some(Rate { limit: 1000, used: 0, remaining: 1000, reset: 1682344037 }), actions_runner_registration: Some(Rate { limit: 10000, used: 0, remaining: 10000, reset: 1682344037 }) }, 
rate: Rate { limit: 5000, used: 0, remaining: 5000, reset: 1682344037 } }
```

As you can see, for search, it actually shows  `0` request used, and 30 remaining (I would have expected to see at least 1 request used as I had a successful request, but maybe it was more than 1 min ago).


I will have a look at how to add retrying based on `x-ratelimit` and try to make a PR...

**Comment by XAMPPRocky** [comment]
@programLyrique I actually forgot that someone did implement this already but I haven't accept the PR 😅 

https://github.com/XAMPPRocky/octocrab/pull/310

**Comment by jrozner** [comment]
I have some code that I wrote that used this API and simply added increasing wait times up until 1 minute in between page requests and was never able to get it to succeed after the first page. It always hit the secondary rate limit. Have you been able to get it to succeed? I can't imagine needing to wait longer than 1 minute.

**Comment by programLyrique** [comment]
> @programLyrique I actually forgot that someone did implement this already but I haven't accept the PR sweat_smile
> 
> #310

Yes, I  actually saw it and had found a feature flag now with some retry capability that can be specified with `add_retry_config` but had not seen the new comments on that PR.

I think the best would be to use the code that extracts the headers of that PR, and add them in the retry of that file. 
I will wait to see if @goodspark  does it!

**Comment by programLyrique** [comment]
> I have some code that I wrote that used this API and simply added increasing wait times up until 1 minute in between page requests and was never able to get it to succeed after the first page. It always hit the secondary rate limit. Have you been able to get it to succeed? I can't imagine needing to wait longer than 1 minute.

The ratelimit API does not give me usable results as even when the error is that I used all the resources, I always have the maximum number of requests remaining.  So I just implemented an exponential backoff with an arbitrary base duration (30s) I think, and usually, after 2 or 3 retries, it worked. I also always waited 1s between every request. The rate limit is over 1h so you might have to wait for 1h if you do too many requests in a short burst.

**Comment by goodspark** [comment]
Search is a bit different. It resets a lot faster (depending on if your request is authenticated). https://docs.github.com/en/rest/search?apiVersion=2022-11-28#rate-limit So it makes sense it would work after 2-3 retries if you start at 30s (bc generally you'll be over 1min by then).

**Comment by Swiddis** [comment]
After running into this I was looking over the [best practices for GitHub's REST API](https://docs.github.com/en/rest/using-the-rest-api/best-practices-for-using-the-rest-api?apiVersion=2022-11-28) and saw that, among other things, they say to avoid making concurrent requests. I suspect, then, that the robust solution here would be to make a (opt-in?) global wrapper that would behave just like the current Octocrab client to callers (and perhaps enforce best practices), but internally queue requests in a `mpsc::Channel` and use the rate limit headers to determine when to next pull from the queue.

The main issue with implementing that as a caller directly is that we don't get access to these headers from the outside for most methods.

**Comment by nataliejameson** [comment]
Just FYI, at least some of this capability was merged in yesterday into the retry layer.
